### PR TITLE
support TypeScript version 4.8

### LINKED ISSUE #1502: Typescript 4.8 support
### Issue workflow progress

_Progress of the issue based on the [Contributor Workflow](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md#a-typical-contributor-workflow)_

- [x] 1. The issue provides a [minimal reproduction](https://en.wikipedia.org/wiki/Minimal_reproducible_example) available on [Stackblitz](https://stackblitz.com/fork/node).
  - _Please install the latest `@envelop/*` packages that you are using._
  - _Please make sure the reproduction is as small as possible._
- [ ] 2. A failing test has been provided
- [ ] 3. A local solution has been provided
- [ ] 4. A pull request is pending review

---

**Describe the bug**

Project's using ts@>4.8.* will not compile when skipLibCheck is disabled:

```
node_modules/@envelop/core/typings/enable-if.d.ts:13:93 - error TS2344: Type 'PluginContextType' does not satisfy the constraint 'Record<string, any>'.

13 export declare function enableIf<PluginContextType = {}>(condition: boolean, plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)): PluginOrDisabledPlugin;
                                                                                               ~~~~~~~~~~~~~~~~~

  node_modules/@envelop/core/typings/enable-if.d.ts:13:34
    13 export declare function enableIf<PluginContextType = {}>(condition: boolean, plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)): PluginOrDisabledPlugin;
                                        ~~~~~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends Record<string, any>` constraint.

node_modules/@envelop/core/typings/enable-if.d.ts:13:128 - error TS2344: Type 'PluginContextType' does not satisfy the constraint 'Record<string, any>'.

13 export declare function enableIf<PluginContextType = {}>(condition: boolean, plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)): PluginOrDisabledPlugin;
                                                                                                                                  ~~~~~~~~~~~~~~~~~

  node_modules/@envelop/core/typings/enable-if.d.ts:13:34
    13 export declare function enableIf<PluginContextType = {}>(condition: boolean, plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)): PluginOrDisabledPlugin;
                                        ~~~~~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends Record<string, any>` constraint.

node_modules/@envelop/core/typings/plugins/use-error-handler.d.ts:4:91 - error TS2344: Type 'ContextType' does not satisfy the constraint 'Record<string, any>'.

4 export declare const useErrorHandler: <ContextType>(errorHandler: ErrorHandler) => Plugin<ContextType>;
                                                                                            ~~~~~~~~~~~

  node_modules/@envelop/core/typings/plugins/use-error-handler.d.ts:4:40
    4 export declare const useErrorHandler: <ContextType>(errorHandler: ErrorHandler) => Plugin<ContextType>;
                                             ~~~~~~~~~~~
    This type parameter might need an `extends Record<string, any>` constraint.
```

**To Reproduce**

install typescript 4.8 in a project using envelop
set `skipLibCheck` to false
run a typescript build

repro: https://github.com/hayes/envelop-ts-4.8-bug

**Environment:**

- OS:
- NodeJS:
- `@envelop/*` versions:
  - `@envelop/core`: `???`

**Additional context**

### PR DESCRIPTION
**Pull request by n1ru4l**
Closes https://github.com/n1ru4l/envelop/issues/1502

### CODE REVIEW

File: packages/plugins/execute-subscription-event/src/index.ts [hunk 1]
```diff
@@ -18,7 +18,7 @@ export type ContextFactoryType<TContextValue = DefaultContext> = (
   options: ContextFactoryOptions
 ) => PromiseOrValue<ContextFactoryHook<TContextValue> | void>;
 
-export const useExtendContextValuePerExecuteSubscriptionEvent = <TContextValue = unknown>(
+export const useExtendContextValuePerExecuteSubscriptionEvent = <TContextValue extends Record<any, any>>(
```
Comment (n1ru4l) [inline packages/plugins/execute-subscription-event/src/index.ts:21]: technically this is a breaking change 🤔

File: packages/core/src/plugins/use-error-handler.ts [hunk 2]
```diff
@@ -10,14 +10,16 @@ type ErrorHandlerCallback<ContextType> = {
...
-  <ContextType>(errorHandler: ErrorHandler) =>
+  <ContextType extends Record<any, any>>(errorHandler: ErrorHandler) =>
   ({ result, args }: ErrorHandlerCallback<ContextType>) => {
     if (result.errors?.length) {
       errorHandler(result.errors, args);
     }
   };
 
-export const useErrorHandler = <ContextType>(errorHandler: ErrorHandler): Plugin<ContextType> => {
+export const useErrorHandler = <ContextType extends Record<string, any>>(
```
Comment (n1ru4l) [inline packages/core/src/plugins/use-error-handler.ts:20]: technically this is a breaking change 🤔

File: packages/core/src/enable-if.ts [hunk 3]
```diff
@@ -16,7 +16,7 @@ export function isPluginEnabled(t: PluginOrDisabledPlugin): t is Plugin {
 /**
  * Utility function to enable a plugin.
  */
-export function enableIf<PluginContextType = {}>(
+export function enableIf<PluginContextType extends Record<any, any> = {}>(
```
Comment (n1ru4l) [inline packages/core/src/enable-if.ts:19]: technically this is a breaking change 🤔

Comment (n1ru4l) [reply packages/core/src/enable-if.ts:19]: but could also be considered as a bug fix, as the earlier type did not satisfy the usage constraint...

File: packages/core/src/plugins/use-error-handler.ts (hunk 2 above)
Comment (n1ru4l) [reply packages/core/src/plugins/use-error-handler.ts:20]: but could also be considered as a bug fix, as the earlier type did not satisfy the usage constraint...

File: packages/plugins/execute-subscription-event/src/index.ts (hunk 1 above)
Comment (n1ru4l) [reply packages/plugins/execute-subscription-event/src/index.ts:21]: but could also be considered as a bug fix, as the earlier type did not satisfy the usage constraint...

**Review by ardatan — APPROVED** [review]

**Review by dotansimha — APPROVED** [review]

**Pull request merged**